appinventor/components/src/com/google/appinventor/components/runtime/File.java

```diff
@@ -16,64 +21,118 @@
 import com.google.appinventor.components.annotations.SimpleFunction;
 import com.google.appinventor.components.annotations.SimpleObject;
 import com.google.appinventor.components.annotations.SimpleProperty;
-import com.google.appinventor.components.annotations.UsesLibraries;
 import com.google.appinventor.components.annotations.UsesPermissions;
+
 import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.common.FileScope;
 import com.google.appinventor.components.common.PropertyTypeConstants;
 import com.google.appinventor.components.common.YaVersion;
-import com.google.appinventor.components.runtime.errors.PermissionException;
+
 import com.google.appinventor.components.runtime.util.AsynchUtil;
+import com.google.appinventor.components.runtime.util.Continuation;
 import com.google.appinventor.components.runtime.util.ErrorMessages;
+import com.google.appinventor.components.runtime.util.FileAccessMode;
+import com.google.appinventor.components.runtime.util.FileOperation;
+import com.google.appinventor.components.runtime.util.FileStreamReadOperation;
+import com.google.appinventor.components.runtime.util.FileStreamWriteOperation;
 import com.google.appinventor.components.runtime.util.FileUtil;
-import com.google.appinventor.components.runtime.util.MediaUtil;
-import com.google.appinventor.components.runtime.util.QUtil;
+import com.google.appinventor.components.runtime.util.FileWriteOperation;
+import com.google.appinventor.components.runtime.util.IOUtils;
+import com.google.appinventor.components.runtime.util.ScopedFile;
+import com.google.appinventor.components.runtime.util.SingleFileOperation;
+import com.google.appinventor.components.runtime.util.Synchronizer;
 
 import java.io.FileNotFoundException;
-import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.InputStreamReader;
+import java.io.OutputStream;
 import java.io.OutputStreamWriter;
-import java.io.StringWriter;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 
 /**
  * Non-visible component for storing and retrieving files. Use this component to write or read files
- * on the device. The default behavior is to write files to the private data directory associated
- * with the app. The Companion writes files to `/sdcard/AppInventor/data` for easy debugging. If
- * the file path starts with a slash (`/`), then the file is created relative to `/sdcard`.
- * For example, writing a file to `/myFile.txt` will write the file in `/sdcard/myFile.txt`.
+ * on the device. File names can take one of three forms:
+ *
+ * - Private files have no leading `/` and are written to app private storage (e.g., "file.txt")
+ * - External files have a single leading `/` and are written to public storage (e.g., "/file.txt")
+ * - Bundled app assets have two leading `//` and can only be read (e.g., "//file.txt")
```

Comment:
I am not getting this. If the `File Scope` is set to `Asset`, then does the file name require to be prefixed by `//`?

Reply:
No, it does not require the `//`. In any case, if you use the `//` prefix then the scope will be switched to `Asset`. I will think about how to refine this to be more clear.